src/nsrt_learning.py

```diff
@@ -146,9 +142,157 @@ def segment_trajectory(trajectory: GroundAtomTrajectory) -> List[Segment]:
+        # Only demonstration trajectories are currently annotated with final
+        # relevant atoms.
+        if not ll_traj.is_demo:
+            continue
+        init_atoms.append(segs[0].init_atoms)
+        final_relevant_atoms.append(ll_traj.goal)
+        skeleton = _find_skeleton(segs, ops_without_sides, option_specs)
+        skeletons.append(skeleton)
+
+    # Convert effects to side predicates.
+    name_to_strips_op = {op.name: op for op in ops_without_sides}
+    # We need to remember the original strips operators because some of the
+    # parameters might get lost during pruning. TODO: there's probably a more
+    # clear way to code this.
+    name_to_original_strips_op = name_to_strips_op.copy()
+    for op, (_, option_vars) in zip(ops_without_sides, option_specs):
+        # Consider converting each add effect.
+        # TODO refactor to avoid redundant code.
+        for effect in op.add_effects:
+            if verbose:
+                print(f"Considering add effect: {effect} from {op.name}")
+            current_op = name_to_strips_op[op.name]
+            assert current_op.name == op.name
+            # Tentatively replace the current operator.
+            remaining_params = {p for atom in current_op.preconditions | \
+                (current_op.add_effects - {effect}) | \
+                current_op.delete_effects
+                for p in atom.variables} | set(option_vars)
+            next_params = [p for p in op.parameters if p in remaining_params]
+            name_to_strips_op[op.name] = STRIPSOperator(
+                op.name, next_params,
+                current_op.preconditions,
+                current_op.add_effects - {effect},
+                current_op.delete_effects,
+                current_op.side_predicates | {effect.predicate})
+            # Check if operators would still cover skeletons.
+            if not all(_skeleton_covered(skeleton, inits, finals,
+                                         name_to_strips_op,
+                                         name_to_original_strips_op)
+                       for (skeleton, inits, finals) in \
+                        zip(skeletons, init_atoms, final_relevant_atoms)):
+                # If not, revert the change.
+                name_to_strips_op[op.name] = current_op
+                if verbose:
+                    print("Skeletons not covered; reverting conversion.")
+            elif verbose:
+                print("Skeletons still covered; keeping conversion.")
+        # Consider converting each delete effect.
+        for effect in op.delete_effects:
+            if verbose:
+                print(f"Considering delete effect: {effect} from {op.name}")
+            current_op = name_to_strips_op[op.name]
+            assert current_op.name == op.name
+            # Tentatively replace the current operator.
+            remaining_params = {p for atom in current_op.preconditions | \
+                (current_op.add_effects - {effect}) | \
```

Comment:
this was a bug right? this should be removed from delete_effects

Reply:
yeah good catch